brave/src/main/java/brave/propagation/ThreadLocalSpan.java

```diff
@@ -0,0 +1,98 @@
+ *
+ * <p>Many libraries expose a callback model as opposed to an interceptor one. When creating new
+ * instrumentation, you may find places where you need to place a span in scope in one callback
+ * (like `onStart()`) and end the scope in another callback (like `onFinish()`).
+ *
+ * <p>Provided the library guarantees these run on the same thread, you can simply propagate the
+ * result of {@link Tracer#withSpanInScope(Span)} from the starting callback to the closing one.
+ * This is typically done with a request-scoped attribute.
+ *
+ * Here's an example:
+ * <pre>{@code
+ * class MyFilter extends Filter {
+ *   public void onStart(Request request, Attributes attributes) {
+ *     // Assume you have code to start the span and add relevant tags...
+ *
+ *     // We now set the span in scope so that any code between here and
+ *     // the end of the request can see it with Tracer.currentSpan()
+ *     SpanInScope spanInScope = tracer.withSpanInScope(span);
+ *
+ *     // We don't want to leak the scope, so we place it somewhere we can
+ *     // lookup later
+ *     attributes.put(SpanInScope.class, spanInScope);
+ *   }
+ *
+ *   public void onFinish(Response response, Attributes attributes) {
+ *     // as long as we are on the same thread, we can read the span started above
+ *     Span span = tracer.currentSpan();
+ *
+ *     // Assume you have code to complete the span
+ *
+ *     // We now remove the scope (which implicitly detaches it from the span)
+ *     attributes.remove(SpanInScope.class).close();
+ *   }
+ * }
+ * }</pre>
+ *
+ * <p>Sometimes you have to instrument a library where There's no attribute namespace shared across
+ * request and response. For this scenario, you can use {@link ThreadLocalSpan} to temporarily store
+ * the span between callbacks.
+ *
+ * Here's an example:
+ * <pre>{@code
+ * class MyFilter extends Filter {
+ *   final ThreadLocalSpan threadLocalSpan;
+ *
+ *   public void onStart(Request request) {
+ *     // Assume you have code to start the span and add relevant tags...
+ *
+ *     // We now set the span in scope so that any code between here and
+ *     // the end of the request can see it with Tracer.currentSpan()
+ *     threadLocalSpan.set(span);
+ *   }
+ *
+ *   public void onFinish(Response response, Attributes attributes) {
+ *     // as long as we are on the same thread, we can read the span started above
+ *     Span span = threadLocalSpan.remove();
+ *     if (span == null) return;
+ *
+ *     // Assume you have code to complete the span
+ *   }
+ * }
+ * }</pre>
+ */
+public final class ThreadLocalSpan {
+  public static ThreadLocalSpan create(Tracer tracer) {
+    return new ThreadLocalSpan(tracer);
+  }
+
+  final Tracer tracer;
+  final ThreadLocal<Tracer.SpanInScope> currentSpanInScope = new ThreadLocal<>();
+
+  ThreadLocalSpan(Tracer tracer) {
+    this.tracer = tracer;
+  }
+
+  public void set(Span span) {
+    currentSpanInScope.set(tracer.withSpanInScope(span));
```

Comment:
I think what will happen is BraveActiveSpanScope will have no idea about this type :) To test, you could in BraveActiveSpanScope's test put a span in scope opentracing doesn't know about, and close that scope (maybe interleaved between ActiveSpan doing the same).

Sometime soon, I'll make a base test class for CurrentTraceContext and put a test in for double-scoping to make sure it doesn't die.